### PR TITLE
add DI container support

### PR DESCRIPTION
**Pull request by LordSimal**
This PR adds DI container support to tasks.

To do this you have to add the 
```
use \Cake\Queue\Queue\ServicesTrait;
```
to your task.

With that you have access to the
```
$service = $this->getService(TestService::class);
```
method from which you can retrieve objects present inside the DI container.

Its basically the same structure/behavior as its currently available in https://github.com/dereuromark/cakephp-queue/pull/327

### CODE REVIEW

**Comment by othercorey** [comment]
Wouldn't DI support inject the services into the task through the constructor?

**Comment by LordSimal** [comment]
I'd have to look into that if we want to build it that way.

**Comment by ndm2** [comment]
I haven't used the Queue plugin yet, but generally, if DI support is being expanded, I'd like to see CakePHP favor injection wherever possible.

While having proper service location is nice, my worry would be that going down that route could make a switch/upgrade to injection harder in the future.

**Comment by LordSimal** [comment]
I only used dereuromarks queue plugin till now and I am happy with how the DI services are being accessible via the trait method mentioned above.

But for the core plugins I can definitely see enforcing constructor injected services as a "rule".

**Comment by LordSimal** [comment]
This now changed to how we are handling service injection into commands.
So services are injectable via the jobs constructor and the definition needs to be present inside the `services()` method of the app/plugin to
* add the job to the DI container and
* add the service as a param to the constructor

But I am not really sure if this can lead to weird results if the worker is running for a longer time of period and the same type of job is being pushed to the queue over and over...

**Comment by ndm2** [comment]
> But I am not really sure if this can lead to weird results if the worker is running for a longer time of period and the same type of job is being pushed to the queue over and over...

Do you have some specific problem in mind?

**Comment by markstory** [comment]
> > But I am not really sure if this can lead to weird results if the worker is running for a longer time of period and the same type of job is being pushed to the queue over and over...
>
> Do you have some specific problem in mind?

One scenario would be services that maintain state and rely on the implicit end of request to 'cleanup' the possibility of shared state.

My position on this is that we should document that services used in workers are long lived and shouldn't retain state. At the end of the day application developers are responsible for using services in queue workers safely.

In the future if this becomes a pain point we can address it then and find a way to reset the container instances after each job is processed.

File: src/Plugin.php [hunk 1]
```diff
@@ -84,4 +85,18 @@ public function console(CommandCollection $commands): CommandCollection
...
+
+    /**
+     * Add DI container to Worker command
+     *
+     * @param \Cake\Core\ContainerInterface $container The DI container
+     * @return void
+     */
+    public function services(ContainerInterface $container): void
+    {
+        $container->add(ContainerInterface::class, $container);
```
Comment (markstory) [inline src/Plugin.php:97]: Should we do this in CakePHP itself as well? This seems like an elegant solution for other framework/plugin abstractions to leverage the DI container.

**Review by markstory — COMMENTED** [review]
Looks good to me. Nice work :clap: I can take care of adding docs for this once it has merged.

**Comment by LordSimal** [comment]
> One scenario would be services that maintain state and rely on the implicit end of request to 'cleanup' the possibility of shared state.

Thinking once more about it: Doesn't the DI Container create new instances by default whenever you ask for a specific service? Only if someone defines it as a shared service inside the `services()` method it will then be a singleton and therefore have state for each job/message process run.

But other shared state could definitely mess up how users expect it work.

File: src/Plugin.php (hunk 1 above)
Comment (LordSimal) [reply src/Plugin.php:97]: yea, this would be a nice default 😄

**Comment by markstory** [comment]
> Only if someone defines it as a shared service inside the services() method it will then be a singleton and therefore have state for each job/message process run.

Yes, someone will do it eventually :smile:

**Pull request merged**